[AutoComplete] Non-unique keys in Menu children when dataSource with duplicated values provided

When I provide `dataSource` values in form of array with objects instead of strings with duplicated `text` values, for example:

```
[
  { text: 'Some Text', value: <span>{someFirstValue}</span> },
  { text: 'Some Text', value: <span>{someSecondValue}</span> }
]
```

`AutoComplete` component renders only one item in `Menu` because Menu children keys are not unique. It also logs a React error with `flattenChildren(...)` warning.

I know that providing duplicated `text` values to `dataSource` is not something normal but I believe that it should be possible. For example I discovered this bug when trying to implement `AutoComplete` for searching for Facebook fanpages, there could be several fanpages with the same name and I would like to show them all in `Menu` and provide some additional context for helping user to find desired fanpage. By additional context I mean something like id, photo, maybe link to fanpage.

I've tracked down fragment of `AutoComplete` component code that causes this error [link](https://github.com/callemall/material-ui/blob/master/src/auto-complete.jsx#L387):

```
if (typeof request.text === 'string') {
  return React.cloneElement(request.value, {
    key: request.text,
    disableFocusRipple: this.props.disableFocusRipple,
  });
}
```

And I would be more than happy to provide a PR with fix.

@straku Now it is possible to do this if you give them unique keys!
```
[
  { key:'uniquekey1', text: 'Some Text', value: <span>{someFirstValue}</span> },
  { key:'uniquekey2', text: 'Some Text', value: <span>{someSecondValue}</span> }
]
```
@mbrookes this issue should have been resolved. however I got 2 linting errors that I have no idea where they from.
```
src/auto-complete.jsx
  340:20  error  Missing trailing comma  comma-dangle
  365:7   error  Missing semicolon       semi
```

--- a/docs/src/app/components/pages/components/AutoComplete/ExampleDataSources.jsx
+++ b/docs/src/app/components/pages/components/AutoComplete/ExampleDataSources.jsx
@@ -25,6 +25,11 @@ const dataSource1 = [
 
 const dataSource2 = ['12345', '23456', '34567'];
 
+const dataSource3 = [
+  {text: 'Some Text', value: 'someFirstValue'},
+  {text: 'Some Text', value: 'someSecondValue'},
+];
+
 const AutoCompleteExampleNoFilter = () => (
   <div>
     <AutoComplete
@@ -37,6 +42,12 @@ const AutoCompleteExampleNoFilter = () => (
       filter={AutoComplete.noFilter}
       openOnFocus={true}
       dataSource={dataSource2}
+    /><br />
+    <AutoComplete
+      floatingLabelText="Same text, different values"
+      filter={AutoComplete.noFilter}
+      openOnFocus={true}
+      dataSource={dataSource3}
     />
   </div>
 );

--- a/src/auto-complete.jsx
+++ b/src/auto-complete.jsx
@@ -287,20 +287,10 @@ const AutoComplete = React.createClass({
 
   handleItemTouchTap(event, child) {
     const dataSource = this.props.dataSource;
-    let chosenRequest;
-    let index;
-    let searchText;
 
-    if (typeof dataSource[0] === 'string') {
-      chosenRequest = this.requestsList[parseInt(child.key, 10)];
-      index = dataSource.indexOf(chosenRequest);
-      searchText = dataSource[index];
-    } else {
-      chosenRequest = child.key;
-      index = dataSource.indexOf(
-        dataSource.filter((item) => chosenRequest === item.text)[0]);
-      searchText = chosenRequest;
-    }
+    const index = parseInt(child.key, 10);
+    const chosenRequest = dataSource[index];
+    const searchText = typeof chosenRequest === 'string' ? chosenRequest : chosenRequest.text;
 
     this.props.onNewRequest(chosenRequest, index);
 
@@ -433,18 +423,48 @@ const AutoComplete = React.createClass({
 
     const requestsList = [];
 
-    dataSource.every((item) => {
+    dataSource.every((item, index) => {
       switch (typeof item) {
         case 'string':
           if (this.props.filter(searchText, item, item)) {
-            requestsList.push(item);
+            requestsList.push({
+              text: item,
+              value: (
+                <MenuItem
+                  innerDivStyle={styles.innerDiv}
+                  value={item}
+                  primaryText={item}
+                  disableFocusRipple={disableFocusRipple}
+                  key={index}
+                />),
+            });
           }
           break;
 
         case 'object':
           if (item && typeof item.text === 'string') {
             if (this.props.filter(searchText, item.text, item)) {
-              requestsList.push(item);
+              if (item.value.type && (item.value.type.displayName === MenuItem.displayName ||
+                 item.value.type.displayName === Divider.displayName)) {
+                requestsList.push({
+                  text: item.text,
+                  value: React.cloneElement(item.value, {
+                    key: index,
+                    disableFocusRipple: this.props.disableFocusRipple,
+                  }),
+                });
+              } else {
+                requestsList.push({
+                  text: item.text,
+                  value: (
+                    <MenuItem
+                      innerDivStyle={styles.innerDiv}
+                      primaryText={item.value}
+                      disableFocusRipple={disableFocusRipple}
+                      key={index}
+                    />),
+                });
+              }
             }
           }
           break;
@@ -467,37 +487,7 @@ const AutoComplete = React.createClass({
         listStyle={Object.assign(styles.list, listStyle)}
         style={Object.assign(styles.menu, menuStyle)}
       >
-        {
-          requestsList.map((request, index) => {
-            switch (typeof request) {
-              case 'string':
-                return (
-                  <MenuItem
-                    disableFocusRipple={disableFocusRipple}
-                    innerDivStyle={styles.innerDiv}
-                    key={index}
-                    value={request}
-                    primaryText={request}
-                  />
-                );
-
-              case 'object':
-                if (typeof request.text === 'string') {
-                  return React.cloneElement(request.value, {
-                    key: request.text,
-                    disableFocusRipple: disableFocusRipple,
-                  });
-                }
-
-                return React.cloneElement(request, {
-                  disableFocusRipple: disableFocusRipple,
-                });
-
-              default:
-                return null;
-            }
-          })
-        }
+        {requestsList.map((i) => i.value)}
       </Menu>
     );
 

